text/0000-oracles.md

```diff
@@ -0,0 +1,117 @@
+
+To make the process more secure trigger owner can grant the asset access only to a special Oracle account known beforehand.
+
+Described solution can be simplified in some edge cases. For example there can be an off-chain woker that always transmits recent data (i.e. currencies). In that case there is no need in a special `OracleRequest`, trigger can read the info directly from the asset.
+
+### Implementing example scenarios
+
+Scenario 1 (quick request):
+
+1. End-user transfers some amount of asset A to a predefined technical account;
+2. Trigger which listens for incoming asset transfers to its technical account reads the metadata of assets where two oracles consistently write data. It reads gold price in dollars and asset A price in dollars. After doing some calculations trigger transfers computed amount of synthetic gold to the end-user account.
+
+Scenario 2 (publisher-consumer):
+
+1. Off-chain Oracle worker receive messages about hurricane from sensors
+2. Worker calls `ExecuteTrigger` instruction on-chain to start benefits transferring
+
+Scenario 3 (request-response):
+
+1. End-user calls trigger T1 using existing `ExecuteTrigger` instruction
+2. Trigger T1 calls `EmitOracleRequest` instruction, providing payload with data like oracle id, request, user data, asset id to write response into and etc. That's all what this trigger does.
+3. Off-chain oracle worker receives the `OracleRequest` event, retrieves user data, makes a request to a University "X" server, and writes the response to an asset specified in original request.
+4. Trigger T2 being triggered on that asset modification. It reads the response from the asset, identifies if end-user is actually a student of University "X" and then grants a permission if so.
+
+
+
+So technically Iroha 2 already supports first two scenarios. This document aims to find a solution for the third scenario.
+
+### Decisions
+
+List all decisions you and other stakeholders made
+
+### Alternatives
+
+1. Make one peer (i.e. *Leader*) directly call some external *API* when trigger submits `EmitOracleRequest` instruction. This is a more on-chain variant and allows us to create integrations with popular oracle providers on our side. However it' not so secure and performant.
+2. Implement our own IBC, make Oracle providers to run their own Iroha 2 based networks and make triggers use this IBC to retrieve data from Oracle blockchains. How exactly Oracle blockchains get data from off-chain world is out of the scope.
+
+### Concerns
+
+There are some concerns and unresolved problems:
+
+1. Payment. Typically Oracles want you to pay them. This is not a big deal for centralized solutions, because in that case payment can be done off chain. But for decentralized automatic solutions we should think how to attach payment to every `OracleRequest` event.
```

Comment:
> ...to attach payment to every `OracleRequest` event.

And, I guess, a robust way to ensure that the request was actually consumed by the oracle.

Reply:
I think this is described by the second point under this section. So that you can be sure that data is really from the exact oracle if it's signed